packages/terra-worklist-data-grid/src/WorklistDataGrid.jsx

```diff
@@ -320,31 +335,94 @@ function WorklistDataGrid(props) {
+    if (rangeInfo.anchorRow > rangeInfo.previousSelectionEndRow) {
+      // The range extends upward from the anchor row
+      if (newEndRowIndex > rangeInfo.previousSelectionEndRow) {
+        // The range was moved down towards the anchor so rows that no longer qualify for the range need to be unselected.
+        rowIdsToUnselect = rows.slice(rangeInfo.previousSelectionEndRow - 1, Math.min(newEndRowIndex, rangeInfo.anchorRow)).map(row => row.id);
+      }
+    } else if (rangeInfo.anchorRow < rangeInfo.previousSelectionEndRow) {
+      // The range extends downward from the anchor row
+      if (newEndRowIndex < rangeInfo.previousSelectionEndRow) {
+        // The range was moved up towards the anchor so rows that no longer qualify for the range need to be unselected. New endRangeIndex becomes ordered End
+        rowIdsToUnselect = rows.slice(Math.max(rangeInfo.anchorRow, newEndRowIndex), rangeInfo.previousSelectionEndRow).map(row => row.id);
+      }
+    }
+
+    const ariaMessage = `${intl.formatMessage({
+      id: hasSelectableRows
+        ? 'Terra.worklist-data-grid.row-selection-multiple-rows-selected'
+        : 'Terra.worklist-data-grid.row-selection-mode-enabled',
+    })}. ${intl.formatMessage({
+      id: rowIdsToSelect.length === 1
+        ? 'Terra.worklist-data-grid.row-selection-template'
+        : 'Terra.worklist-data-grid.row-selection-selected-rows-range',
+    }, { row: rangeInfo.anchorRow, endRow: newEndRowIndex })}`;
+
+    setAriaLiveMessage(ariaMessage);
+    onRowSelect(rowIdsToSelect, rowIdsToUnselect);
   };
 
   const selectRows = (selectAllRows, rowId, rowIndex) => {
     let rowLabel;
+    let isSelectAction = true;
     let msgId = 'Terra.worklist-data-grid.all-rows-selected';
+    // Reset last selected row when all rows are selected.
+    rowSelection.current.rangeUsingLastIndividuallySelectedRow = null;
     if (!selectAllRows) {
-      const isSelectAction = !rows[rowIndex - 1].isSelected; // Determine if this is select or unselected.
-      msgId = isSelectAction ? 'Terra.worklist-data-grid.row-selection-template' : 'Terra.worklist-data-grid.row-selection-cleared-template';
+      isSelectAction = !rows[rowIndex - 1].isSelected; // Determine if this is select or unselected.
+      // Remember the last selected row
+      rowSelection.current.rangeUsingLastIndividuallySelectedRow = { anchorRow: rowIndex };
+      if (isSelectAction) {
+        msgId = 'Terra.worklist-data-grid.row-selection-template';
+      } else {
+        msgId = 'Terra.worklist-data-grid.row-selection-cleared-template';
+      }
       rowLabel = rows[rowIndex - 1].ariaLabel || (rowIndex + 1);
     }
     setAriaLiveMessage(intl.formatMessage({ id: msgId }, { row: rowLabel }));
     if (selectAllRows && onRowSelectAll) {
       onRowSelectAll();
     } else if (onRowSelect) {
-      onRowSelect(rowId);
+      if (isSelectAction) {
+        onRowSelect([rowId]);
+      } else {
+        onRowSelect([], [rowId]);
```

Comment:
Why are we passing an empty array for the first argument?

Reply:
onRowSelect accepts two arguments to explicitly communicate the user's intention. The first being rows they would like to select and the second are the rows they would like to unselect.